webapp/graphite/tags/redis.py

```diff
@@ -27,68 +29,119 @@ def __init__(self):
 
+  @logtime()
   def find_series(self, tags):
-    with self.r.pipeline() as pipe:
-      all_match_empty = True
-
-      for tagspec in tags:
-        (tag, operator, spec) = self.parse_tagspec(tagspec)
-
-        # find list of values that match the tagspec
-        values = None
-
-        if operator == '=':
-          matches_empty = spec == ''
-
-          values = [spec]
-
-        elif operator == '=~':
-          pattern = re.compile(spec)
-          matches_empty = bool(pattern.match(''))
-
-          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if pattern.match(value) is not None]
-
-        elif operator == '!=':
-          matches_empty = spec != ''
-
-          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if value != spec]
-
-        elif operator == '!=~':
-          pattern = re.compile(spec)
-          matches_empty = not pattern.match('')
-
-          values = [value for value in self.r.sscan_iter('tags:' + tag + ':values') if pattern.match(value) is None]
-
-        else:
-          raise ValueError("Invalid operator %s" % operator)
-
-        # if we're matching the empty value, union with the set of series that don't have the tag
-        if matches_empty:
-          # store series that don't have this tag in a temp key
-          pipe.sdiffstore('temp:' + tag + ':empty', 'series', 'tags:' + tag + ':series')
-
-          # store union of series without this tag and all series that match the spec in a temp key
-          pipe.sunionstore('temp:' + tagspec, 1 + len(values), 'temp:' + tag + ':empty', *['tags:' + tag + ':values:' + value for value in values])
-        # otherwise only return series for matching values
-        else:
-          # if this query matched no values, just short-circuit since the result of the final intersect will be empty
-          if not values:
-            return []
-
-          # store union of all series that match the spec in a temp key
-          pipe.sunionstore('temp:' + tagspec, len(values), *['tags:' + tag + ':values:' + value for value in values])
-
-        all_match_empty = all_match_empty and matches_empty
-
-      if all_match_empty:
-        raise ValueError("At least one tagspec must not match the empty string")
-
-      # get the final result from the intersection of the temp lists we created above
-      pipe.sinter(*['temp:' + tagspec for tagspec in tags])
-
-      results = pipe.execute()
-
-    return sorted(list(results[-1]))
+    selector = None
+    selector_cnt = None
+    filters = []
+
+    # loop through tagspecs, look for best = or =~ match
+    for tagspec in tags:
+      (tag, operator, spec) = self.parse_tagspec(tagspec)
+
+      if operator == '=':
+        matches_empty = spec == ''
+        if not matches_empty and (not selector or selector[1] != '='):
+          cnt = self.r.scard('tags:' + tag + ':values:' + spec)
+          if not selector or selector_cnt > cnt:
```

Comment:
If the first expression evaluated has the `=~` operator it will be used as the selector and the number of values for its key will be stored as `selector_cnt`. If then a `=` expression comes next, the number of values of this `key=value` combo will be compared against the number of values that the key of the selector refers to. I think even if the number of values the key of the `=~` refers to is smaller than the number of ids the `=` expression points, the `=` is still preferable.